File: phoenix-rise-sdk/python/phoenix_sdk/models/trader.py
from __future__ import annotations

from enum import Enum

from .core import Decimal, PhoenixModel, Side
from .market import RiskState, RiskTier


class TraderActivityState(str, Enum):
    UNINITIALIZED = "uninitialized"
    COLD = "cold"
    ACTIVE = "active"
    REDUCE_ONLY = "reduceOnly"
    FROZEN = "frozen"


class CapabilityAccess(PhoenixModel):
    immediate: bool = False
    via_cold_activation: bool = False


class TraderCapabilitiesView(PhoenixModel):
    place_limit_order: CapabilityAccess
    place_market_order: CapabilityAccess
    risk_increasing_trade: CapabilityAccess
    risk_reducing_trade: CapabilityAccess
    deposit_collateral: CapabilityAccess
    withdraw_collateral: CapabilityAccess


class TraderPositionView(PhoenixModel):
    symbol: str
    position_size: Decimal
    virtual_quote_position: Decimal
    entry_price: Decimal
    unrealized_pnl: Decimal
    discounted_unrealized_pnl: Decimal
    position_initial_margin: Decimal
    initial_margin: Decimal
    maintenance_margin: Decimal
    backstop_margin: Decimal
    limit_order_margin: Decimal
    position_value: Decimal
    unsettled_funding: Decimal
    accumulated_funding: Decimal
    liquidation_price: Decimal
    take_profit_price: Decimal | None = None
    stop_loss_price: Decimal | None = None


class LimitOrder(PhoenixModel):
    price: Decimal
    side: Side
    order_sequence_number: str
    initial_trade_size: Decimal
    trade_size_remaining: Decimal
    margin_requirement: Decimal
    margin_factor: Decimal
    is_reduce_only: bool
    is_stop_loss: bool = False


class TraderView(PhoenixModel):
    flags: int
    state: TraderActivityState
    capabilities: TraderCapabilitiesView
    trader_key: str
    trader_pda_index: int
    trader_subaccount_index: int
    authority: str
    collateral_balance: Decimal
    effective_collateral: Decimal
    effective_collateral_for_withdrawals: Decimal
    unrealized_pnl: Decimal
    discounted_unrealized_pnl: Decimal
    unsettled_funding_owed: Decimal
    accumulated_funding: Decimal
    portfolio_value: Decimal
    maintenance_margin: Decimal
    cancel_margin: Decimal
    initial_margin: Decimal
    initial_margin_for_withdrawals: Decimal
    risk_state: RiskState
    risk_tier: RiskTier
    positions: list[TraderPositionView]
    limit_orders: dict[str, list[LimitOrder]]
    maker_fee_override_multiplier: float
    taker_fee_override_multiplier: float
    max_positions: int
    last_deposit_slot: int
    is_in_active_traders: bool


class TraderStateResponse(PhoenixModel):
    slot: int
    slot_index: int
    authority: str
    pda_index: int
    traders: list[TraderView]

File: ember-frontend/src/types/trader.ts
// SDK decimal value — all numeric fields use this format
export interface SdkDecimalValue {
  decimals: number;
  ui: string;
  value: number;
}

// Margin mode for positions — cross uses shared collateral, isolated uses dedicated collateral
export type MarginMode = "cross" | "isolated";

// Display-ready position (post-transform from SDK's camelCase + Decimal format).
// SDK sends: positionSize (Decimal, signed), entryPrice (Decimal), etc.
// traderStore.transformPosition() converts to this flat format.
export interface TraderPosition {
  symbol: string;
  side: string; // "Long" | "Short"
  size: number;
  entry_price: number;
  mark_price: number;
  unrealized_pnl: number;
  discounted_unrealized_pnl: number;
  margin_mode: MarginMode;
  allocated_collateral: number;
  liquidation_price: number | null;
  position_value: number;
  initial_margin: number;
  // Funding (separate from mark-to-market unrealized PnL).
  // unsettled_funding: this epoch, not yet paid; jumps to ~0 at the 8h boundary
  //   when settlement folds it into accumulated_funding.
  // accumulated_funding: lifetime funding for this position (signed; negative
  //   = paid). Used in cumulative PnL accounting; should NOT be added to the
  //   headline mark-to-market unrealized PnL.
  unsettled_funding: number;
  accumulated_funding: number;
  tp_price: number | null;
  sl_price: number | null;
  subaccount_index: number; // 0 = cross-margin, 1-100 = isolated
}

// Display-ready limit order (post-transform from SDK's camelCase + Decimal/string format).
// SDK sends: initialTradeSize (Decimal), tradeSizeRemaining (Decimal),
// orderSequenceNumber (string), priceTicks (string from WS).
// traderStore.transformLimitOrder() converts to this flat format.
export interface LimitOrder {
  price_in_ticks: number;
  order_sequence_number: number;
  side: string; // "Bid" | "Ask"
  price: number;
  size: number;
  remaining_size: number;
  subaccount_index: number; // 0 = cross-margin, 1-100 = isolated
}

// SDK TraderView account — matches actual API response (camelCase)
export interface TraderAccount {
  traderKey: string;
  traderPdaIndex: number;
  traderSubaccountIndex: number;
  authority: string;
  state: string;
  flags: number; // Phoenix capability bitmask; 63 = fully activated
  collateralBalance: SdkDecimalValue;
  effectiveCollateral: SdkDecimalValue;
  portfolioValue: SdkDecimalValue;
  unrealizedPnl: SdkDecimalValue;
  discountedUnrealizedPnl: SdkDecimalValue;
  maintenanceMargin: SdkDecimalValue;
  cancelMargin: SdkDecimalValue;
  initialMargin: SdkDecimalValue;
  riskState: string;
  riskTier: string;
  capabilities?: Record<string, { immediate: boolean; viaColdActivation: boolean }>;
  positions: TraderPosition[];
  limitOrders: Record<string, LimitOrder[]>;
}

// GET /api/trader/:pubkey response
export interface TraderResponse {
  authority: string;
  accounts: TraderAccount[];
}

// Cancel order ID format sent to backend.
// NOTE: backend expects price as USD (e.g. 50.0) — it converts to ticks server-side.
export interface CancelOrderId {
  price: number;
  order_sequence_number: number;
}

// Order history item from SDK — camelCase, string values
export interface OrderHistoryItem {
  orderSequenceNumber: number;
  marketSymbol: string;
  status: string;
  side: string;
  isReduceOnly: boolean;
  price: string;
  baseQty: string;
  remainingBaseQty: string;
  filledBaseQty: string;
  placedAt: string;
  completedAt: string | null;
}

// Trade history item from SDK — camelCase, string values
export interface TradeHistoryItem {
  marketSymbol: string;
  baseQty: string;
  quoteQty: string;
  price: string;
  timestamp: string; // ISO 8601 string, e.g. "2026-02-12T03:36:48Z"
  transactionSignature: string;
  instructionType: string;
}

// Funding history event from SDK
export interface FundingHistoryEvent {
  timestamp: number;
  symbol: string;
  funding_payment: string;
  funding_rate_percentage: number;
  position_size: number;
  position_side: string;
}
